### ISSUE TITLE
Style/RedundantStringEscape has bad interaction with squiggly heredoc

### ISSUE THREAD

**Issue by amake** [rubocop/rubocop#11346]
I have a unit test where the expected value is a squiggly heredoc with a leading space like so:

```ruby
assert_equal(<<~TXT.chomp, segments.last)
  \ very long string that has a leading space
TXT
```

The Style/RedundantStringEscape cop flags the escaped leading space, and wants to autocorrect it to remove the escape. The Layout/HeredocIndentation cop then wants to remove the space, resulting in a non-equivalent string.

I tried to use an interpolated string literal (`#{' '}very long string...`) but this is flagged by Lint/LiteralInInterpolation.

I know there are various other contortions I could try to build the desired string, but that seems like it shouldn't be necessary.

--------

## Expected behavior

The Style/RedundantStringEscape shouldn't flag an escape necessary for the meaning of the string, and shouldn't remove it as an autofix.

## Actual behavior

Style/RedundantStringEscape will in some cases flag an escape that is necessary for the meaning of the string, and will remove it as an autofix.

## Steps to reproduce the problem

Write a squiggly heredoc containing a leading escaped whitespace character.

## RuboCop version

```
$ bundle exec rubocop -V
1.41.1 (using Parser 3.1.3.0, rubocop-ast 1.24.0, running on ruby 2.7.7) [x86_64-darwin22]
  - rubocop-performance 1.15.2
```
